File: sdks/typescript/src/providers/slh-dsa-192f.ts
/**
 * SLH-DSA-SHA2-192f CryptoProvider — wraps @noble/post-quantum.
 *
 * FIPS 205, NIST Level 3. Backup PQ signature (hash-based, conservative).
 * Public key: 48 B, Signature: 35664 B.
 * Slow signing (~160 ms) but mathematically independent from lattice-based ML-DSA.
 */

import { slh_dsa_sha2_192f } from '@noble/post-quantum/slh-dsa.js';
import type { CryptoProvider } from '../core/crypto-provider.js';
import { deriveKid } from '../core/crypto-provider.js';

const BASE64 = {
  encode(bytes: Uint8Array): string {
    return Buffer.from(bytes).toString('base64');
  },
  decode(str: string): Uint8Array {
    return new Uint8Array(Buffer.from(str, 'base64'));
  },
};

export class SlhDsa192fProvider implements CryptoProvider {
  readonly alg = 'slh-dsa-192f';
  readonly keySize = 48;
  readonly sigSize = 35664;
  readonly isConstantTime = true;

  async generateKeypair(): Promise<{
    kid: string;
    publicKeyB64: string;
    secretKeyB64: string;
  }> {
    const keys = slh_dsa_sha2_192f.keygen();
    const kid = deriveKid(this.alg, keys.publicKey);
    return {
      kid,
      publicKeyB64: BASE64.encode(keys.publicKey),
      secretKeyB64: BASE64.encode(keys.secretKey),
    };
  }

  async sign(
    message: Uint8Array,
    secretKeyB64: string,
  ): Promise<Uint8Array> {
    const sk = BASE64.decode(secretKeyB64);
    return slh_dsa_sha2_192f.sign(message, sk);
  }

  async verify(
    message: Uint8Array,
    signature: Uint8Array,
    publicKeyB64: string,
  ): Promise<boolean> {
    const pk = BASE64.decode(publicKeyB64);
    try {
      return slh_dsa_sha2_192f.verify(signature, message, pk);
    } catch {
      return false;
    }
  }
}

File: sdks/rust/src/providers/slh_dsa_192f.rs
use base64::Engine;
use base64::engine::general_purpose::STANDARD as BASE64;
use fips205::slh_dsa_sha2_192f;
use fips205::traits::{KeyGen, SerDes, Signer, Verifier};
use zeroize::Zeroize;

use crate::v2::crypto_provider::{CryptoError, CryptoProvider, GeneratedKeypair, derive_kid};

/// SLH-DSA-SHA2-192f provider (FIPS 205, NIST Level 3).
/// Backup PQ family — hash-based, conservative, mathematically independent from ML-DSA.
pub struct SlhDsa192fProvider;

const SK_LEN: usize = 96;
const PK_LEN: usize = 48;
const SIG_LEN: usize = 35664;

impl CryptoProvider for SlhDsa192fProvider {
    fn alg(&self) -> &str {
        "slh-dsa-192f"
    }

    fn key_size(&self) -> usize {
        PK_LEN
    }

    fn sig_size(&self) -> usize {
        SIG_LEN
    }

    fn is_constant_time(&self) -> bool {
        true
    }

    fn generate_keypair(&self) -> Result<GeneratedKeypair, CryptoError> {
        let (pk, sk) = slh_dsa_sha2_192f::KG::try_keygen()
            .map_err(|e| CryptoError::KeyError(format!("SLH-DSA-192f keygen failed: {:?}", e)))?;
        let pk_bytes = pk.into_bytes();
        let mut sk_bytes = sk.into_bytes();
        let kid = derive_kid(self.alg(), &pk_bytes);

        let result = GeneratedKeypair {
            kid,
            public_key_b64: BASE64.encode(&pk_bytes),
            secret_key_b64: BASE64.encode(&sk_bytes),
        };

        sk_bytes.zeroize();
        Ok(result)
    }

    fn sign(&self, message: &[u8], secret_key_b64: &str) -> Result<Vec<u8>, CryptoError> {
        let mut sk_bytes = BASE64.decode(secret_key_b64)
            .map_err(|e| CryptoError::KeyError(format!("base64 decode: {}", e)))?;
        let sk_arr: &[u8; SK_LEN] = sk_bytes.as_slice().try_into()
            .map_err(|_| CryptoError::KeyError(format!("invalid SLH-DSA-192f secret key length: expected {}", SK_LEN)))?;
        let sk = slh_dsa_sha2_192f::PrivateKey::try_from_bytes(sk_arr)
            .map_err(|e| CryptoError::KeyError(format!("invalid SLH-DSA-192f secret key: {:?}", e)))?;
        sk_bytes.zeroize();

        let sig = sk.try_sign(message, &[], false)
            .map_err(|e| CryptoError::SignatureError(format!("SLH-DSA-192f sign failed: {:?}", e)))?;
        Ok(sig.to_vec())
    }

    fn verify(&self, message: &[u8], signature: &[u8], public_key_b64: &str) -> Result<bool, CryptoError> {
        let pk_bytes = BASE64.decode(public_key_b64)
            .map_err(|e| CryptoError::KeyError(format!("base64 decode: {}", e)))?;
        let pk_arr: &[u8; PK_LEN] = pk_bytes.as_slice().try_into()
            .map_err(|_| CryptoError::KeyError(format!("invalid SLH-DSA-192f public key length: expected {}", PK_LEN)))?;
        let pk = slh_dsa_sha2_192f::PublicKey::try_from_bytes(pk_arr)
            .map_err(|e| CryptoError::KeyError(format!("invalid SLH-DSA-192f public key: {:?}", e)))?;
        let sig_arr: &[u8; SIG_LEN] = signature.try_into()
            .map_err(|_| CryptoError::SignatureError(format!("invalid SLH-DSA-192f signature length: expected {}", SIG_LEN)))?;
        Ok(pk.verify(message, sig_arr, &[]))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_keypair_generation() {
        let provider = SlhDsa192fProvider;
        let kp = provider.generate_keypair().unwrap();
        assert_eq!(kp.kid.len(), 32);
    }

    #[test]
    fn test_sign_and_verify() {
        let provider = SlhDsa192fProvider;
        let kp = provider.generate_keypair().unwrap();
        let message = b"hello sphincs";
        let sig = provider.sign(message, &kp.secret_key_b64).unwrap();
        let valid = provider.verify(message, &sig, &kp.public_key_b64).unwrap();
        assert!(valid);
    }

    #[test]
    fn test_verify_wrong_message() {
        let provider = SlhDsa192fProvider;
        let kp = provider.generate_keypair().unwrap();
        let sig = provider.sign(b"correct", &kp.secret_key_b64).unwrap();
        let valid = provider.verify(b"wrong", &sig, &kp.public_key_b64).unwrap();
        assert!(!valid);
    }
}
